spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -24,95 +24,103 @@ import geotrellis.vector._
 
 import com.vividsolutions.jts.geom.prep.{PreparedGeometry, PreparedGeometryFactory}
 import org.apache.spark.rdd._
+import scala.util.Try
 
 object ClipToGrid {
   /** Trait which contains methods to be used in determining
     * the most optimal way to clip geometries and features
     * for ClipToGrid methods.
     */
   trait Predicates {
-    /** Returns true if the feature geometry covers the passed in extent */
+    /** True if the feature geometry covers the passed-in [[Extent]]. */
     def covers(e: Extent): Boolean
-    /** Returns true if the feature geometry is covered bythe passed in extent */
+    /** True if the feature geometry is covered by the passed-in [[Extent]]. */
     def coveredBy(e: Extent): Boolean
   }
 
-  /** Clips a feature to the given extent, that uses the given predicates
+  /** Clips a feature to the given [[Extent]], using the given [[Predicates]]
     * to avoid doing intersections where unnecessary.
     */
   def clipFeatureToExtent[G <: Geometry, D](
     e: Extent,
-    feature: Feature[G, D],
+    f: Feature[G, D],
     preds: Predicates
-  ): Option[Feature[Geometry, D]] =
-    if(preds.covers(e)) { Some(Feature(e, feature.data)) }
-    else if(preds.coveredBy(e)) { Some(feature) }
-    else {
-      feature.geom.intersection(e).toGeometry.map { g =>
-        Feature(g, feature.data)
-      }
-    }
+  ): Option[Feature[Geometry, D]] = e match {
+    /* If a Feature covers the Extent, their intersection would be the Extent itself. */
+    case _ if preds.covers(e) => Some(Feature(e, f.data))
+    /* The Feature may be completely contained within the Extent. In that case, no clipping need occur at all. */
+    case _ if preds.coveredBy(e) => Some(f)
+    /* Otherwise, we need to perform a JTS intersection */
+    case _ => Try(f.geom.intersection(e)).toOption.flatMap(_.toGeometry.map(g => Feature(g, f.data)))
```

Comment:
The intersection can still throw Topology Exceptions in weird cases, so we guard against it here.

Reply:
Haha we could debate this one to the moon. I agree in general that swallowing errors isn't good. Throwing exceptions in `IO` code is sometimes correct. Throwing exceptions in pure code is never correct.

Now with soapboxing out of the way... I refactored this bit first before I realized how it was used further down. The analogue to this in VectorPipe also takes a logging function, which logs in the case of `None` returned by any clippers (like this function). What's our path forward?

Reply:
> If there's issues with the input, the user should know.

Yes, we're totally on the same page about this. Swallowing errors is never good (we can leave the exception debate aside). So the GeoTrellis version of this function only returns in `Option` because of the `.toGeometry` then, yeah? Given that there's no logging function passed in to deal with failed clips right away (like in VP), I'm having trouble seeing the integration path with VP. I'll prototype some more on this PR and hopefully we can reach something nice.

Reply:
Logging while being robust and skipping failed clips was a specific requirement of VectorPipe, so that one naughty geom doesn't crash the whole ingest. The only way to do that would be to catch Topology Exceptions directly within an `RDD.map`, so the injected `.clipToGrid` methods here wouldn't be usable by VP as-is.

Still pondering, I'll have an answer some time this morning.